### DISCUSSION TITLE
JSON formats for adding recipe via API

### DISCUSSION THREAD

**Discussion by jecorn** [mealie-recipes/mealie#2264] [Q&A] [answered]
(edit: looks like the code formatting for logs/json doesn't work out very well. I'm open to suggestions to make it more readable)
I love the layout and API-centric nature of mealie. But I'm trying to wrap my head around the JSON format in 1.0.0b. At the bottom of this message is an example of something that I'm trying to add via the post /api/recipes. It took me a while, but I finally figured out that I needed to hash the recipeInstructions to get them to show up.

But now, If I copy/paste this into mealie itself via the JSON editor, then everything looks fine in the GUI. **But** the GUI also displays a warning of "No Entry Found". If I go back to the recipe after saving, then it suddenly becomes the default "empty" recipe template. 

**My overall question:** What in the json scheme is optional vs required on initial recipe post? For example, the image can't be on the filesystem when the JSON is first sent, but should the entry by there? Can recipeIngredient hashes/ids be generated automatically if missing from the text? 

A tail of the docker log is:

`
mealie2  | INFO:     127.0.0.1:44242 - "POST /api/recipes HTTP/1.1" 201 Created
mealie2  | INFO:     127.0.0.1:44244 - "GET /api/recipes/mashed-potatoes-with-scallions-and-horseradish HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:44260 - "GET /api/media/recipes/410f2a5d-2530-4561-8afd-38180d19a7ef/images/original.webp?rnd=1&version= HTTP/1.1" 404 Not Found
mealie2  | INFO:     127.0.0.1:44276 - "GET /api/recipes/mashed-potatoes-with-scallions-and-horseradish/timeline/events?page=1&perPage=-1&orderBy=timestamp&orderDirection=asc HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:44288 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:44302 - "GET /api/recipes/mashed-potatoes-with-scallions-and-horseradish/comments HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:44306 - "GET /api/organizers/tags?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:44310 - "GET /api/organizers/categories?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:58730 - "GET /api/recipes/mashed-potatoes-with-scallions-and-horseradish/comments HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:58732 - "GET /api/app/about HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:58740 - "GET /api/media/recipes/undefined/images/original.webp?rnd=1&version= HTTP/1.1" 404 Not Found
mealie2  | ERROR: 18-Mar-23 21:47:22 	No Entry Found on recipe controller action
mealie2  | INFO:     127.0.0.1:58754 - "PUT /api/recipes/undefined HTTP/1.1" 404 Not Found
mealie2  | INFO:     127.0.0.1:58770 - "GET /api/recipes/timeline/events?page=1&perPage=-1&orderBy=timestamp&orderDirection=asc HTTP/1.1" 404 Not Found
mealie2  | INFO:     127.0.0.1:58768 - "GET /api/organizers/tools?page=1&perPage=-1&orderBy=name&orderDirection=asc HTTP/1.1" 200 OK
mealie2  | INFO:     127.0.0.1:44150 - "GET /api/app/about HTTP/1.1" 200 OK
`

If I try to use the API to send this JSON, then things get even weirder. The request is accepted normally with response 201 and the slug as the response string. A recipe is created with the intended title for the recipe with the right title and slug. But the recipe itself is the default newly created empty recipe.

The offending JSON
`{
    "name": "Mashed Potatoes with Scallions and Horseradish",
    "description": "When it comes to mashed potatoes, most cooks worry so much\nabout getting the texture right that they forget about the flavor. Butter and\nhalf-and-half make for mashed potatoes that are rich tasting but not terribly exciting. Our\ngoal was to jazz up the traditional recipe without muddying the flavor of the\npotatoes, so we used simple pairings of contrasting yet complementary\ningredients. ",
    "totalTime": "",
    "recipeYield": "4",
    "recipeIngredient": [
        {
            "note": "2 pounds russet potatoes, unpeeled and scrubbed"
        },
        {
            "note": "8 tablespoons unsalted butter (1 stick), melted"
        },
        {
            "note": "1 \u00bd teaspoons table salt"
        },
        {
            "note": "\u00bd teaspoon ground black pepper"
        },
        {
            "note": "2 tablespoons prepared horseradish"
        },
        {
            "note": "\u00bc cup grated fresh horseradish"
        },
        {
            "note": "3 medium scallions, green parts only, minced"
        },
        {
            "note": "1 cup half-and-half, warm"
        }
    ],
    "recipeInstructions": [
        {
            "text": "Place potatoes in large saucepan with water to cover by 1 inch. Bring to boil over high heat, reduce heat to medium, and simmer until potatoes are just tender (paring knife can be slipped into and out of potato with very little resistance), 20 to 30 minutes. Drain. ",
            "id": "cda6187d653df7ee7386e115a9f5403b"
        },
        {
            "text": "Set food mill or ricer over now-empty but still-warm saucepan. Spear potato with dinner fork, then peel off skin with paring knife; repeat with remaining potatoes. Working in batches, cut peeled potatoes into rough chunks and drop into hopper of food mill or ricer. Process or rice potatoes into saucepan or mash potatoes with potato masher directly in saucepan. ",
            "id": "502a6e0f999446b7009f59c08f610462"
        },
        {
            "text": "Stir butter into potatoes until just incorporated. Sprinkle salt and pepper over potatoes. Whisk horseradish and scallions into warm half-and-half; add mixture to potatoes and stir until just combined. Serve immediately. ",
            "id": "9fd404937175b23e494afd4cbedc26d2"
        }
    ],
    "org_url": "https://www.americastestkitchen.com/recipes/1-mashed-potatoes-with-scallions-and-horseradish",
    "settings": {
        "public": true,
        "showNutrition": false,
        "showAssets": true,
        "landscapeView": true,
        "disableComments": false,
        "disableAmount": false
    }
}`

**Comment by michael-genson** [comment, marked as answer]
Be sure to check out the API docs at `https://your-url.com/docs` (if you're using the omni image this is currently bugged, you need to navigate to your backend url which is `https://your-ip-address:9000/docs`)

When you create the recipe, you only supply the name. To edit everything else you do a `PUT` (replace all) or `PATCH` (replace some) request. The reason the `POST` returns just the slug (as a raw string) is because you need it for the subsequent update request.

**Reply by jecorn** [reply]
Thanks! I am using omni, which must be why I couldn't find the API docs. I was trying to piece things together from things like https://nightly.mealie.io/api/redoc/#tag/Recipe:-CRUD/operation/create_one_api_recipes_post

Now that I'm posting the name, I noticed that the string returned is `"string"` and not `string`. Which means a bit of extra string stripping. Any reason why the returned string has the extra quotes?

**Reply by michael-genson** [reply]
I think that's just how FastAPI renders the docs. I'm pretty sure the string doesn't have extra quotes. Let me know if you see otherwise, though

**Reply by jecorn** [reply]
The quotes are definitely there in the response text.

`response = requests.post(mealie_url + "/api/recipes", headers=headers, json={"name":data["name"]})`

`print(response.text)`
returns `"mashed-potatoes-with-scallions-and-horseradish"`

while `print(response.text)[1:-1]`
returns `mashed-potatoes-with-scallions-and-horseradish`

a subsequent
`response = requests.put(mealie_url + "/api/recipes/"+response.text, headers=headers, json=data)`
`print(response.text)`
returns `{"detail":{"message":"No Entry Found","error":true,"exception":null}}`
So it's unable to find the entry using the intact string, perhaps due to the quotes.

Oddly though, I'm also getting an error on the `put` if I slice out the quotes:
`response = requests.put(mealie_url + "/api/recipes/"+response.text[1:-1], headers=headers, json=data)`
`print(response.text)` 
returns `{"detail":{"message":"Recipe already exists","error":true,"exception":null}}`
and the recipe contains just the default empty recipe

**Reply by mrfenyx** [reply]
Had the same issue. In the end I figured out I had to use PATCH to only send some details, not replace everything in the recipe.
